packages/ember-metal/lib/expand_properties.js

```diff
@@ -42,31 +42,103 @@ export default function expandProperties(pattern, callback) {
     pattern.indexOf(' ') === -1
   );
 
-  var parts = pattern.split(SPLIT_REGEX);
-  var properties = [parts];
+  var properties = [pattern];
 
-  for (let i = 0; i < parts.length; i++) {
-    let part = parts[i];
-    if (part.indexOf(',') >= 0) {
-      properties = duplicateAndReplace(properties, part.split(','), i);
+  for (let property, i = 0; i < properties.length; i++) {
+    for (let match; (property = properties[i]) && (match = matchOuterMostBraces(property));) {
+      properties.splice(i, 1, ...duplicateAndReplace(property, match));
     }
   }
 
   for (let i = 0; i < properties.length; i++) {
-    callback(properties[i].join('').replace(END_WITH_EACH_REGEX, '.[]'));
+    callback(properties[i].replace(END_WITH_EACH_REGEX, '.[]'));
   }
 }
 
-function duplicateAndReplace(properties, currentParts, index) {
+function duplicateAndReplace(property, match) {
   var all = [];
+  var parts = splitOutsideBraces(match[1], ',');
 
-  properties.forEach((property) => {
-    currentParts.forEach((part) => {
-      var current = property.slice(0);
-      current[index] = part;
-      all.push(current);
-    });
+  parts.forEach((part) => {
```

Comment:
Good point about the map, @stefanpenner.

Since there's an overall desire to reduce function calls, I'll rewrite this as a straight for loop.